### ISSUE TITLE
Public key download blocks event loop and takes the service down

### ISSUE THREAD

**Issue by brunobat** [smallrye/smallrye-jwt#511]
When we provide a URI to download the certs containing the public keys, if that endpoint fails to respond, the thread get's locked and causes issues both on service startup and also unintended service restarts triggered by kubernetes. 

The rest client should be setup with a short and configurable read and connection timeouts.

When we use the library on Quarkus 1.13.4 under Java 11, At least.
    <dependency>
      <groupId>io.quarkus</groupId>
      <artifactId>quarkus-smallrye-jwt</artifactId>
    </dependency>

Stacktrace:

`Oct 4, 2021 @ 10:55:27.689	2021-10-04 10:55:27.688 WARNING [io.ver.cor.imp.BlockedThreadChecker] (vertx-blocked-thread-checker) [spn=:trc=:tid=][aid=:uid=] Thread Thread[vert.x-eventloop-thread-0,5,main]=Thread[vert.x-eventloop-thread-0,5,main] has been blocked for 77274 ms, time limit is 2000 ms: io.vertx.core.VertxException: Thread blocked

	Oct 4, 2021 @ 10:55:27.689		at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$2.handle(HttpSecurityRecorder.java:51)

	Oct 4, 2021 @ 10:55:27.689		at com.nr.vertx.instrumentation.VertxUtil$1.handle(VertxUtil.java:117)

	Oct 4, 2021 @ 10:55:27.689		at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:101)

	Oct 4, 2021 @ 10:55:27.689		at java.base@11.0.5/sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:746)

	Oct 4, 2021 @ 10:55:27.689		at io.quarkus.smallrye.jwt.runtime.auth.MpJwtValidator$1.accept(MpJwtValidator.java:53)

	Oct 4, 2021 @ 10:55:27.689		at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)

	Oct 4, 2021 @ 10:55:27.689		at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)

	Oct 4, 2021 @ 10:55:27.689		at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:36)

	Oct 4, 2021 @ 10:55:27.689		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)

	Oct 4, 2021 @ 10:55:27.689		at io.smallrye.jwt.auth.principal.DefaultJWTTokenParser.parse(DefaultJWTTokenParser.java:56)

	Oct 4, 2021 @ 10:55:27.689		at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)

	Oct 4, 2021 @ 10:55:27.689		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)

	Oct 4, 2021 @ 10:55:27.689		at java.base@11.0.5/java.lang.Thread.run(Thread.java:834)

	Oct 4, 2021 @ 10:55:27.689		at io.smallrye.jwt.auth.principal.KeyLocationResolver.<init>(KeyLocationResolver.java:44)

	Oct 4, 2021 @ 10:55:27.689		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)

	Oct 4, 2021 @ 10:55:27.689		at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)

	Oct 4, 2021 @ 10:55:27.689	
	Oct 4, 2021 @ 10:55:27.689		at java.base@11.0.5/java.net.SocketInputStream.read(SocketInputStream.java:168)

	Oct 4, 2021 @ 10:55:27.689		at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:229)

	Oct 4, 2021 @ 10:55:27.689		at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:324)

	Oct 4, 2021 @ 10:55:27.689		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)

	Oct 4, 2021 @ 10:55:27.689		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)

	Oct 4, 2021 @ 10:55:27.688		at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$2.handle(HttpSecurityRecorder.java:51)

	Oct 4, 2021 @ 10:55:27.688		at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1038)

	Oct 4, 2021 @ 10:55:27.688		at io.vertx.ext.web.impl.RouterImpl.handle(RouterImpl.java:36)

	Oct 4, 2021 @ 10:55:27.688		at io.quarkus.vertx.http.runtime.VertxHttpRecorder$1.handle(VertxHttpRecorder.java:131)

	Oct 4, 2021 @ 10:55:27.688		at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)

	Oct 4, 2021 @ 10:55:27.688		at io.smallrye.mutiny.groups.UniSubscribe.withSubscriber(UniSubscribe.java:50)

	Oct 4, 2021 @ 10:55:27.688		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)

	Oct 4, 2021 @ 10:55:27.688		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)

	Oct 4, 2021 @ 10:55:27.688		at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)

	Oct 4, 2021 @ 10:55:27.688		at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)

	Oct 4, 2021 @ 10:55:27.688		at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)

	Oct 4, 2021 @ 10:55:27.688		at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:137)

	Oct 4, 2021 @ 10:55:27.688		at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)`

**Comment by sberyozkin** [comment]
Hi @brunobat thanks for opening this issue, I think we'll need to fetch the remote keys immediately at the start up, right now it is done at the 1st request time.
I can also  suggest a workaround, create a custom mechanism:

https://quarkus.io/guides/security-customization#httpauthenticationmechanism-customization

All it will do is it will extend [JWTAuthMechanism](https://github.com/quarkusio/quarkus/blob/main/extensions/smallrye-jwt/runtime/src/main/java/io/quarkus/smallrye/jwt/runtime/auth/JWTAuthMechanism.java), and override its `authenticate` method like this:

```
String jwtToken = // simply extract it from `Authorization Bearer`, copy that extractor code if you'd like but unless you support different headers/cookies, it is not needed 
        if (jwtToken != null) {
            return identityProviderManager
                    .authenticateBlocking(new TokenAuthenticationRequest(new JsonWebTokenCredential(jwtToken)));
        }
        return Uni.createFrom().optional(Optional.empty());
```

Note `authenticateBlocking` call - that should fix it for you until the next release.

The other alternative is to investigate using `quarkus-oidc` - however if you do not use the keys from Keycloak or other providers then it won't work

**Comment by sberyozkin** [comment]
@brunobat Good point, I forgot `HttpAuthenticationMechanism` has no blocking context available, please see https://github.com/quarkusio/quarkus/pull/20635 for further discussion related to it.

Re the possible refresh - yes, this is why https://github.com/quarkusio/quarkus/pull/20635 would be the best solution in the short term, while reading the keys early would only be a good optimization.
Also, please check `quarkus-oidc`

**Comment by brunobat** [comment]
Will try it and get back to you. Thanks @sberyozkin !

**Comment by sberyozkin** [comment]
@brunobat https://github.com/quarkusio/quarkus/pull/20635 has been merged so it will make it to in 2.4.0.Final. If you won't be able to migrate then indeed experiment with a custom `IdentityProvider`:
```
@ApplicationScoped
@AltPriority(1)
public class BlockingMpJwtValidator {
@Inject
MpJwtValidator validator;

    @Override
    public Uni<SecurityIdentity> authenticate(TokenAuthenticationRequest request,
            AuthenticationRequestContext context) {
        return context.runBlocking(() -> validator.authenticate(request, context));
    }
}
``` 

should do it.

**Comment by sberyozkin** [comment]
@brunobat I'm closing this issue as it it is really a Quarkus issue which has been worked-around with https://github.com/quarkusio/quarkus/pull/20635.

smallrye-jwt enhancement which can eliminate such issues is #345 

thanks
